internal/controller/nnf_lustre_mgt_controller.go

```diff
@@ -0,0 +1,393 @@
+			},
+		}
+
+		if err := r.Get(ctx, client.ObjectKeyFromObject(configMap), configMap); err != nil {
+			return nil, dwsv1alpha2.NewResourceError("could not get Lustre MGT start fsname config map: %v", client.ObjectKeyFromObject(configMap)).WithError(err).WithMajor()
+		}
+
+		if configMap.Data == nil {
+			configMap.Data = make(map[string]string)
+		}
+		configMap.Data["NextFsName"] = fsname
+
+		if err := r.Update(ctx, configMap); err != nil {
+			if apierrors.IsConflict(err) {
+				return &ctrl.Result{Requeue: true}, nil
+			}
+			return nil, dwsv1alpha2.NewResourceError("could not update Lustre MGT used fsname config map:  %v", client.ObjectKeyFromObject(configMap)).WithError(err).WithMajor()
+		}
+	}
+
+	// Don't set the next fsname in the status section until after the configmap is updated
+	nnfLustreMgt.Status.FsNameNext = fsname
+
+	return nil, nil
+}
+
+// HandleNewClaims looks for any new claims in Spec.ClaimList and assigns them
+// an fsname
+func (r *NnfLustreMGTReconciler) HandleNewClaims(ctx context.Context, nnfLustreMgt *nnfv1alpha1.NnfLustreMGT) (*ctrl.Result, error) {
+	claimMap := map[corev1.ObjectReference]string{}
+	for _, claim := range nnfLustreMgt.Status.ClaimList {
+		claimMap[claim.Reference] = claim.FsName
+	}
+
+	for _, reference := range nnfLustreMgt.Spec.ClaimList {
+		if _, exists := claimMap[reference]; !exists {
+			fsnameNext := nnfLustreMgt.Status.FsNameNext
+
+			result, err := r.SetFsNameNext(ctx, nnfLustreMgt, fsnameNext)
+			if err != nil {
+				return nil, err
+			}
+			if result != nil {
+				return result, nil
+			}
+
+			newClaim := nnfv1alpha1.NnfLustreMGTStatusClaim{
+				Reference: reference,
+				FsName:    fsnameNext,
+			}
+
+			nnfLustreMgt.Status.ClaimList = append(nnfLustreMgt.Status.ClaimList, newClaim)
+
+			return nil, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// RemoveOldClaims removes any old entries from the Status.ClaimList and erases the fsname from
+// the MGT if necessary.
+func (r *NnfLustreMGTReconciler) RemoveOldClaims(ctx context.Context, nnfLustreMgt *nnfv1alpha1.NnfLustreMGT) error {
+	claimMap := map[corev1.ObjectReference]bool{}
+	for _, reference := range nnfLustreMgt.Spec.ClaimList {
+		claimMap[reference] = true
+	}
+
+	for i, claim := range nnfLustreMgt.Status.ClaimList {
+		if _, exists := claimMap[claim.Reference]; !exists {
+			if err := r.EraseOldFsName(nnfLustreMgt, claim.FsName); err != nil {
+				return err
+			}
+
+			nnfLustreMgt.Status.ClaimList = append(nnfLustreMgt.Status.ClaimList[:i], nnfLustreMgt.Status.ClaimList[i+1:]...)
+
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (r *NnfLustreMGTReconciler) EraseOldFsName(nnfLustreMgt *nnfv1alpha1.NnfLustreMGT, fsname string) error {
+	log := r.Log.WithValues("NnfLustreMGT", client.ObjectKeyFromObject(nnfLustreMgt))
+
+	if r.ControllerType == ControllerRabbit {
+		if cmd, found := os.LookupEnv("NNF_LUSTRE_FSNAME_ERASE_COMMAND"); found {
+			log.Info("Erasing fsname from MGT", "fsname", fsname)
```

Comment:
We seem to be conflating MGT with MGS, but it probably doesn't matter.
It occurs to me that almost every use of MGT in this PR is really referring to an MGS rather than an individual MGT.

Reply:
I think MGT makes sense here. That's where all of the fsname data is being stored. There can be multiple MGSs that have access to the MGT, and we wouldn't want to track the MGSs individually since they serve up the same target.